skate-plugin/src/main/kotlin/com/slack/sgp/intellij/featureflags/FeatureFlagAnnotator.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2023 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.sgp.intellij.featureflags
+
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.ide.BrowserUtil
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.ExternalAnnotator
+import com.intellij.lang.annotation.HighlightSeverity
+import com.intellij.openapi.application.runReadAction
+import com.intellij.openapi.components.service
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.slack.sgp.intellij.SkatePluginSettings
+import com.slack.sgp.intellij.TEST_KOTLIN_LANGUAGE_ID_KEY
+import com.slack.sgp.intellij.util.isLinkifiedFeatureFlagsEnabled
+import java.net.URI
+import org.jetbrains.kotlin.idea.KotlinLanguage
+
+class FeatureFlagAnnotator : ExternalAnnotator<PsiFile, List<FeatureFlagSymbol>>() {
+  private val extractor by lazy { FeatureFlagExtractor() }
+
+  override fun collectInformation(file: PsiFile): PsiFile? {
+    if (!isKotlinFeatureFile(file)) {
+      return null
+    }
+    if (extractor.getFeatureFlagsForPsiFile(file) == null) {
+      val flags = extractor.extractFeatureFlags(file)
+      extractor.setFeatureFlagsForPsiFile(file, flags)
+    }
+    return file
+  }
+
+  override fun doAnnotate(collectedInfo: PsiFile): List<FeatureFlagSymbol> {
+    if (!collectedInfo.project.isLinkifiedFeatureFlagsEnabled() || !isKotlinFile(collectedInfo)) {
+      return emptyList()
+    }
+
+    val flagsForFile = extractor.getFeatureFlagsForPsiFile(collectedInfo)
+    return transformToFeatureFlagSymbols(collectedInfo, flagsForFile)
+  }
```

Comment:
I'm curious why both of these functions need to be implemented?

Reply:
As per the documentation 

```
Implemented by a custom language plugin to process the files in a language by an external annotation tool ("linter"). External annotators are expected to be (relatively) slow and are started after regular annotators have completed their work.
Annotators work in three steps:
collectInformation(PsiFile, Editor, boolean) is called to collect some data about a file needed for launching a tool
collected data is passed to doAnnotate which executes a tool and collects highlighting data
highlighting data is applied to a file by apply
```

Based on that my understanding it splits the responsibilities in a clear way. In our case if collectInformation thinks it is not a featureFile then we won't even need to call doAnnotate

Reply:
hmmm, I went and read the [source](https://github.com/JetBrains/intellij-community/blob/master/platform/analysis-api/src/com/intellij/lang/annotation/ExternalAnnotator.java) and you omitted some details. Namely that `collectInformation` is called within a read action and `doAnnotate` is not and should be for long-running tasks.

After reading it there, I believe what you should be doing here is this
- Change the first generic to `List<FeatureFlagSymbol>`. This is computed in `collectInformation` and, when returned in that function, will be passed on to `doAnnotate()`. Currently you use `PsiFile` and just read it once and use it as a cache key in `extractor` in `doAnnotate()`.
- It seems like `!collectedInfo.project.isLinkifiedFeatureFlagsEnabled() || !isKotlinFile(collectedInfo)` should be moved to the `collectInformation()` function.
- `doAnnotate` doesn't need to do any extra processing as far as I can tell, so seems it can return the collected data and leave it on for the `apply()` function to use

With that, you can also remove the cache from `FeatureFlagExtractor` and just make it an `object` class rather than instantiable as it's purely functional in nature.

Reply:
[252b25b](https://github.com/slackhq/slack-gradle-plugin/pull/576/commits/252b25b00fef199674f7b82f9895f3f943dfa6fc)